File: sdks/python/agentcomms/resources/ai.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 Victory (Intergalactic Tech).
# Licensed under the Apache License, Version 2.0. See LICENSE for details.
"""AI resource — Bedrock-backed operations on messages."""
from __future__ import annotations

from typing import TYPE_CHECKING, Any, Optional

if TYPE_CHECKING:
    from agentcomms.client import Client


class AiResource:
    def __init__(self, client: "Client", agent_id: str) -> None:
        self._client = client
        self._agent_id = agent_id

    def _path(self, op: str) -> str:
        return f"/agents/{self._agent_id}/ai/{op}"

    def categorize(
        self,
        *,
        message_id: str,
        labels: Optional[list[str]] = None,
        categories: Optional[list[str]] = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {"message_id": message_id}
        selected_labels = labels or categories
        if selected_labels:
            body["labels"] = selected_labels
        return self._client._request("POST", self._path("categorize"), json=body)

    def extract(
        self,
        *,
        message_id: str,
        schema: dict[str, Any] | None = None,
        fields: Optional[list[str]] = None,
    ) -> dict[str, Any]:
        if schema is None:
            if not fields:
                raise ValueError("schema is required")
            schema = {
                "type": "object",
                "properties": {field: {"type": "string"} for field in fields},
            }
        return self._client._request("POST", self._path("extract"), json={"message_id": message_id, "schema": schema})

    def summarize(
        self,
        *,
        text: str | None = None,
        message_id: str | None = None,
        thread_key: str | None = None,
        length: str | None = None,
        max_length: Optional[int] = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {}
        if text:
            body["text"] = text
        if message_id:
            body["message_id"] = message_id
        if thread_key:
            body["thread_key"] = thread_key
        if length:
            body["length"] = length
        elif max_length:
            body["length"] = "long" if max_length > 500 else "short"
        return self._client._request("POST", self._path("summarize"), json=body)

    def search(self, *, query: str, limit: int = 10) -> dict[str, Any]:
        return self._client._request("POST", self._path("search"), json={"query": query, "limit": limit})

File: sdks/node/src/resources/ai.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Victory (Intergalactic Tech).
// Licensed under the Apache License, Version 2.0. See LICENSE for details.
import type { Client } from "../client.js";

export class AiResource {
  constructor(private client: Client, private agentId: string) {}

  private path(op: string): string {
    return `/agents/${this.agentId}/ai/${op}`;
  }

  async categorize(params: { message_id: string; labels?: string[]; categories?: string[] }): Promise<Record<string, unknown>> {
    return this.client.request("POST", this.path("categorize"), {
      message_id: params.message_id,
      labels: params.labels ?? params.categories,
    });
  }

  async extract(params: { message_id: string; schema?: Record<string, unknown>; fields?: string[] }): Promise<Record<string, unknown>> {
    const schema =
      params.schema ??
      (params.fields
        ? {
            type: "object",
            properties: Object.fromEntries(params.fields.map((field) => [field, { type: "string" }])),
          }
        : undefined);
    if (!schema) throw new Error("schema is required");
    return this.client.request("POST", this.path("extract"), {
      message_id: params.message_id,
      schema,
    });
  }

  async summarize(params: { text?: string; message_id?: string; thread_key?: string; length?: "short" | "long"; max_length?: number }): Promise<{ summary: string }> {
    return this.client.request("POST", this.path("summarize"), {
      text: params.text,
      message_id: params.message_id,
      thread_key: params.thread_key,
      length: params.length ?? (params.max_length && params.max_length > 500 ? "long" : undefined),
    });
  }

  async search(params: { query: string; limit?: number }): Promise<{ results: unknown[] }> {
    return this.client.request("POST", this.path("search"), params);
  }
}
